Kernel+LibC: Signals galore

IdanHo opened the pull request.

bgianfo commented:
Does `sigwaitinfo()` need to be it's own separate syscall? It looks like it's normally implemented in just LibC as a wrapper around `sigtimedwait()`?

For example in MUSL: https://git.musl-libc.org/cgit/musl/tree/src/signal/sigwaitinfo.c

```c
#include <signal.h>

int sigwaitinfo(const sigset_t *restrict mask, siginfo_t *restrict si)
{
	return sigtimedwait(mask, si, 0);
}
```

IdanHo commented:
Posix says: `If timeout is the null pointer, the behavior is unspecified.`, so not sure if that's allowed?

bgianfo commented:
> Posix says: `If timeout is the null pointer, the behavior is unspecified.`, so not sure if that's allowed?

But it's our implementation.. so we can specify the behavior how ever we want? :thinking:

It just means that there is no guarantee to external callers, right?

IdanHo commented:
> But it's our implementation.. so we can specify the behavior how ever we want? thinking
> 
> It just means that there is no guarantee to external callers, right?

If that's the intended meaning then sure, I just wasn't sure if they meant that it's undefined behaviour of the C/C++ kind, will merge them then

bgianfo reviewed (approved):
LGTM!

IdanHo commented on Kernel/Syscalls/sigaction.cpp:
```diff
@@ -307,4 +307,22 @@ ErrorOr<FlatPtr> Process::sys$sigaltstack(Userspace<const stack_t*> ss, Userspac
     return 0;
 }
 
+ErrorOr<FlatPtr> Process::sys$sigwaitinfo(Userspace<sigset_t const*> set, Userspace<siginfo_t*> info)
```
I think they should be added to the location which contains the majority of the implementation, so for sigwait I would put it in LibC, while in the functions which are just thin wrappers around a syscall I would put it in the Kernel, will add it for all 3 now :)

IdanHo commented:
Updated sys$sigtimedwait to treat nullptr timeout as no timeout, and replaced the sigwaitinfo libc function to be a thin wrapper around sys$sigtimedwait.
